dhall/src/Dhall/Import.hs

```diff
@@ -1019,62 +1000,105 @@ getCacheBaseDirectory = alternative₀ <|> alternative₁ <|> alternative₂
 
 -- If the URL contains headers typecheck them and replace them with their normal
 -- forms.
-normalizeHeaders :: URL -> StateT Status IO URL
-normalizeHeaders url@URL { headers = Just headersExpression } = do
+normalizeHeadersIn :: URL -> StateT Status IO URL
+normalizeHeadersIn url@URL { headers = Just headersExpression } = do
     Status { _stack } <- State.get
     loadedExpr <- loadWith headersExpression
+    let handler (e :: SomeException) = throwMissingImport (Imported _stack e)
+    normalized <- liftIO $ handle handler (normalizeHeaders loadedExpr)
+    return url { headers = Just (fmap absurd normalized) }
+
+normalizeHeadersIn url = return url
+
+-- | Empty origin headers used for remote contexts
+--   (and fallback when nothing is set in env or config file)
+emptyOriginHeaders :: Expr Src Import
+emptyOriginHeaders = ListLit (Just (fmap absurd originHeadersTypeExpr)) mempty
+
+-- | A fake Src to annotate headers expressions with.
+--   We need to wrap headers expressions in a Note for nice error reporting,
+--   and because `?` handling only catches SourcedExceptions
+headersSrc :: Src
+headersSrc = Src {
+        srcStart = SourcePos {
+            sourceName = fakeSrcName,
+            sourceLine = mkPos 1,
+            sourceColumn = mkPos 1
+        },
+        srcEnd = SourcePos {
+            sourceName = fakeSrcName,
+            sourceLine = mkPos 1,
+            sourceColumn = mkPos (Text.length fakeSrcText)
+        },
+        srcText = fakeSrcText
+    }
+  where
+    fakeSrcText = "«Origin Header Configuration»"
+    fakeSrcName = "[builtin]"
+
+-- | Load headers only from the environment (used in tests)
+envOriginHeaders :: Expr Src Import
+envOriginHeaders = Note headersSrc (Embed (Import (ImportHashed Nothing (Env "DHALL_HEADERS")) Code))
+
+-- | Load headers in env, falling back to config file
+defaultOriginHeaders :: IO (Expr Src Import)
+defaultOriginHeaders = do
+    fromFile <- originHeadersFileExpr
+    return (Note headersSrc (ImportAlt envOriginHeaders (Note headersSrc fromFile)))
+
+-- | Given a headers expression, return an origin headers loader
+originHeadersLoader :: IO (Expr Src Import) -> StateT Status IO OriginHeaders
+originHeadersLoader headersExpr = do
+
+    -- Load the headers using a parallel state with an empty impport chain.
+    -- We also set _loadOriginHeaders to prevent reentrant loads.
+
+    status <- State.get
+
+    let headerLoadStatus = status {
+        _stack = pure (NonEmpty.last (_stack status)),
```

Comment:
If we change this to `_stack = headerImport :| NonEmpty.tail (_stack status)` then it will do the right thing and not trigger the referential sanity check.  Then I think you wouldn't need the `reentrantLoad` work-around.

That does, however, mean that you would need to get the `headerImport` which is currently not in scope, but I'll leave a separate comment for that.